### PR TITLE
[AnnData setup refactor] Make setup_anndata a static method on model classes rather than one global function

### LINKED ISSUE #1107: Docstring reuse
Let's use this tool [`docrep`](https://docrep.readthedocs.io/en/latest/?badge=latest) to reduce duplicate code in the form of reused docstrings. See [cellrank](https://cellrank.readthedocs.io/en/stable/api.html) for an [example](https://github.com/theislab/cellrank/blob/master/cellrank/ul/_docs.py).

E.g., 

https://github.com/theislab/cellrank/blob/395bdccaa3509594818d5ce99d030cb9394d4b52/cellrank/tl/kernels/_velocity_kernel.py#L54-L69

### LINKED ISSUE #1135: [AnnData setup refactor] Make setup_anndata a static class method on the model
See [this](https://docs.google.com/document/d/18howmTv4eg6dOBIEaJTpUr1DsS2GaaYmZewgy6lB_ME/edit?usp=sharing) doc for details

Note: We originally considered calling the static setup_anndata method from respective model constructors. However we later decided not to do so due to at least the two issues below:
1. This would be a breaking change as it would break users' flow: today users set up anndata then create the model per our user guides, but with this change the new setup_anndata call with default parameters would clear out their original setup.
2. Ceratin models have factory methods to create them from other models (see SCANVI for example). In these cases we need to provide a way to not set up the anndata again as we intent to use the anndata that was already set up in the original model.

### PR DESCRIPTION
**Pull request by watiss**
Closes #1135 
Closes #1107

### CODE REVIEW

File: scvi/model/_scanvi.py [hunk 1]
```diff
@@ -172,7 +172,7 @@ def from_scvi_model(
         unlabeled_category
             Value used for unlabeled cells in `labels_key` used to setup AnnData with scvi.
         adata
-            AnnData object that has been registered via :func:`~scvi.data.setup_anndata`.
+            AnnData object that has been registered via :meth:`~scvi.model.SCANVI.setup_anndata`.
```
Comment (watiss) [inline scvi/model/_scanvi.py:175]: @adamgayoso  Is this correct, or is it SCVI (or any other model)?

File: scvi/external/gimvi/_model.py [hunk 2]
```diff
@@ -519,6 +520,61 @@ def load(
         model.to_device(device)
         return model
 
+    @staticmethod
+    def setup_anndata(
+        adata: AnnData,
```
Comment (watiss) [inline scvi/external/gimvi/_model.py:525]: @adamgayoso Is this correct?

File: scvi/external/solo/_model.py [hunk 3]
```diff
@@ -388,6 +388,47 @@ def auto_forward(module, x):
 
         return preds_df
 
+    @staticmethod
+    def setup_anndata(
+        adata: AnnData,
+        layer: Optional[str] = None,
+        copy: bool = False,
```
Comment (watiss) [inline scvi/external/solo/_model.py:395]: @adamgayoso Do we need this for this model?

File: scvi/external/gimvi/_model.py [hunk 4]
```diff
@@ -519,6 +520,61 @@ def load(
...
+        batch_key: Optional[str] = None,
+        labels_key: Optional[str] = None,
+        categorical_covariate_keys: Optional[List[str]] = None,
+        continuous_covariate_keys: Optional[List[str]] = None,
+        copy: bool = False,
```
Comment (watiss) [inline scvi/external/gimvi/_model.py:530]: @adamgayoso Are the params here correct?

File: scvi/external/gimvi/_model.py [hunk 5]
```diff
@@ -519,6 +520,61 @@ def load(
...
+        Sets up the :class:`~anndata.AnnData` object for this model.
+
+        A mapping will be created between data fields used by this model to their respective locations in adata.
+        This method will also compute the log mean and log variance per batch for the library size prior.
+
+        None of the data in adata are modified. Only adds fields to adata.
+
+        Parameters
+        ----------
+        adata
```
Comment (jjhong922) [inline scvi/external/gimvi/_model.py:542]: Resuable documentation would be really useful for these. Relevant notion task: https://www.notion.so/See-if-towncrier-can-automate-release-notes-92866f9e1e6b48218b4690b3041370b7

File: scvi/external/solo/_model.py (hunk 3 above)
Comment (jjhong922) [reply scvi/external/solo/_model.py:395]: I think copy should stay on all the methods, since its really just a general convenience function

File: scvi/model/_condscvi.py [hunk 6]
```diff
@@ -202,3 +203,49 @@ def train(
             plan_kwargs=plan_kwargs,
             **kwargs,
         )
+
+    @staticmethod
+    def setup_anndata(
+        adata: AnnData,
+        labels_key: Optional[str] = None,
```
Comment (jjhong922) [inline scvi/model/_condscvi.py:210]: I think for a lot of these model specific setup_anndata methods, we can be more strict on what arguments are required now. E.g. in this case, the model doesn't really make sense without a labels_key. @adamgayoso

File: scvi/model/_scanvi.py (hunk 1 above)
Comment (jjhong922) [reply scvi/model/_scanvi.py:175]: pretty sure its SCANVI here

File: scvi/model/base/_base_model.py [hunk 7]
```diff
@@ -32,9 +32,7 @@ class BaseModelClass(ABC):
     def __init__(self, adata: Optional[AnnData] = None):
         if adata is not None:
             if "_scvi" not in adata.uns.keys():
-                raise ValueError(
-                    "Please setup your AnnData with scvi.data.setup_anndata(adata) first"
-                )
+                raise ValueError("Please set up your AnnData with setup_anndata first")
```
Comment (jjhong922) [inline scvi/model/base/_base_model.py:35]: Would it be possible to do something like __name__.setup_anndata() to customize the message for the model class?

File: tests/models/test_models.py [hunk 8]
```diff
@@ -166,7 +172,7 @@ def test_scvi(save_path):
     batch = np.zeros(a.n_obs)
     batch[:64] += 1
     a.obs["batch"] = batch
-    setup_anndata(a, batch_key="batch")
+    _setup_anndata(a, batch_key="batch")
```
Comment (jjhong922) [inline tests/models/test_models.py:175]: should this be SCVI.setup_anndata?, same with the ones below

File: tests/models/test_scarches.py [hunk 9]
```diff
@@ -163,7 +163,7 @@ def test_scanvi_online_update(save_path):
     new_labels = adata1.obs.labels.to_numpy()
     new_labels[0] = "Unknown"
     adata1.obs["labels"] = pd.Categorical(new_labels)
-    setup_anndata(adata1, batch_key="batch", labels_key="labels")
+    _setup_anndata(adata1, batch_key="batch", labels_key="labels")
```
Comment (jjhong922) [inline tests/models/test_scarches.py:166]: should this be SCANVI.setup_anndata (same with below)

**Review by jjhong922 — COMMENTED** [review]
overall looks good, left nits and some comments on how we can make stuff clearer for the users.

File: scvi/data/_anndata.py [hunk 10]
```diff
@@ -10,6 +10,7 @@
 import pandas as pd
 import rich
 from anndata._core.anndata import AnnData
+from deprecate import deprecated
```
[the remark this answers is not in the document]
Comment (watiss) [reply scvi/data/_anndata.py:13]: https://borda.github.io/pyDeprecate/

File: tests/models/test_models.py (hunk 8 above)
Comment (watiss) [reply tests/models/test_models.py:175]: I didnt want to descrease coverage with the _ setup_anndata since we still have it around. So I oped to use the model-specific setup_anndata call in some spots, and the _ setup_anndata call in other spots

File: tests/models/test_scarches.py (hunk 9 above)
Comment (watiss) [reply tests/models/test_scarches.py:166]: resolving as dupe of the one on another test

File: scvi/_docs.py [hunk 11]
```diff
@@ -37,7 +37,7 @@
 batch_correction
     Whether to correct for batch effects in DE inference.
 batchid1
-    Subset of categories from `batch_key` registered in :func:`~scvi.data.setup_anndata`,
+    Subset of categories from `batch_key` registered in setup_anndata,
```
[the remark this answers is not in the document]
Comment (watiss) [reply scvi/_docs.py:40]: I looked at https://www.sphinx-doc.org/en/1.4.9/markup/inline.html but since we arent specifying a :role: here I'm not sure the advantage would be?

File: scvi/data/_anndata.py [hunk 12]
```diff
@@ -30,7 +31,7 @@ def get_from_registry(adata: anndata.AnnData, key: str) -> np.ndarray:
     Parameters
     ----------
     adata
-        anndata object already setup with `scvi.data.setup_anndata()`
+        anndata object already setup with setup_anndata
```
[the remark this answers is not in the document]
Comment (watiss) [reply scvi/data/_anndata.py:34]: resolving as dupe of the one above, will address all if we decide we need it

File: scvi/dataloaders/_concat_dataloader.py [hunk 13]
```diff
@@ -15,7 +15,7 @@ class ConcatDataLoader(DataLoader):
     Parameters
     ----------
     adata
-        AnnData object that have been registered via :func:`~scvi.data.setup_anndata`.
+        AnnData object that has been registered via setup_anndata.
```
[the remark this answers is not in the document]
Comment (watiss) [reply scvi/dataloaders/_concat_dataloader.py:18]: resolving as dupe of the one above. Will find/replace all instances if we decide we need it in the other coment

File: scvi/model/_condscvi.py (hunk 6 above)
Comment (adamgayoso) [reply scvi/model/_condscvi.py:210]: agree, same with Solo labels key

File: scvi/model/_destvi.py [hunk 14]
```diff
@@ -358,3 +359,49 @@ def train(
             plan_kwargs=plan_kwargs,
             **kwargs,
         )
+
+    @staticmethod
+    def setup_anndata(
+        adata: AnnData,
+        labels_key: Optional[str] = None,
```
Comment (adamgayoso) [inline scvi/model/_destvi.py:366]: I'm not sure if this model needs the labels key, I believe not @romain-lopez

File: scvi/model/_linear_scvi.py [hunk 15]
```diff
@@ -125,3 +127,49 @@ def get_loadings(self) -> pd.DataFrame:
         )
 
         return loadings
+
+    @staticmethod
+    def setup_anndata(
+        adata: AnnData,
+        batch_key: Optional[str] = None,
```
Comment (adamgayoso) [inline scvi/model/_linear_scvi.py:134]: I believe this one needs the labels key as well as it has a dispersion option of gene-label

File: scvi/model/base/_archesmixin.py [hunk 16]
```diff
@@ -41,7 +41,7 @@ def load_query_data(
         ----------
         adata
             AnnData organized in the same way as data used to train model.
-            It is not necessary to run :func:`~scvi.data.setup_anndata`,
+            It is not necessary to run setup_anndata,
```
Comment (adamgayoso) [inline scvi/model/base/_archesmixin.py:44]:
```suggestion
            It is not necessary to run `setup_anndata`,
```

This happens in several places, the ` just make it format in italics on docs

File: tests/external/test_solo.py [hunk 17]
```diff
@@ -1,12 +1,13 @@
...
+from scvi.data._anndata import _setup_anndata
 from scvi.external import SOLO
 from scvi.model import SCVI
 
 
 def test_solo(save_path):
     n_latent = 5
     adata = synthetic_iid(run_setup_anndata=False)
-    setup_anndata(adata)
+    _setup_anndata(adata)
```
Comment (adamgayoso) [inline tests/external/test_solo.py:10]: all the tests should call the appropriate static method?

File: tests/dataset/test_built_in_data.py [hunk 18]
```diff
@@ -20,29 +21,29 @@ def test_populate(self):
...
-        scvi.data.setup_anndata(gene_dataset)
+        _setup_anndata(gene_dataset)
         unsupervised_training_one_epoch(gene_dataset)
 
 
 class TestSeqfishDataset(TestCase):
     def test_populate(self):
         dataset = scvi.data.seqfish(save_path="tests/data")
-        scvi.data.setup_anndata(dataset)
+        _setup_anndata(dataset)
```
Comment (adamgayoso) [inline tests/dataset/test_built_in_data.py:46]: we can probably add setup anndata to the `unsupervised_training_one_epoch` function? and then remove all these lines

File: tests/models/test_models.py (hunk 8 above)
Comment (adamgayoso) [reply tests/models/test_models.py:175]: would it decrease coverage if it's just calling it behind the scenes? (sorry it's late during this review :) )

**Review by adamgayoso — CHANGES_REQUESTED** [review]
Looks good,  a few minor things. I'm wondering if we can build a decorator to automatically build the setup anndata docstring with the option to add text to each param

File: scvi/_docs.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jjhong922) [reply scvi/_docs.py:40]: We should still have it use the `code` formatting. For rst its actually double back ticks. See: https://docs.scvi-tools.org/en/stable/api/reference/scvi.model.LinearSCVI.html for an example

File: tests/models/test_models.py (hunk 8 above)
Comment (jjhong922) [reply tests/models/test_models.py:175]: it shouldn't decrease coverage I believe, code_cov steps through the entire execution

File: scvi/model/base/_archesmixin.py (hunk 16 above)
Comment (jjhong922) [reply scvi/model/base/_archesmixin.py:44]: do we want to use code formatting instead (double back ticks)

File: scvi/external/gimvi/_model.py (hunk 5 above)
Comment (watiss) [reply scvi/external/gimvi/_model.py:542]: agree but the documentation is not exactly the same here contrary to what is the case with doc_differential_expression @adamgayoso

File: scvi/model/_condscvi.py (hunk 6 above)
Comment (watiss) [reply scvi/model/_condscvi.py:210]: are you sure the model doesn't really makes sense without a labels_key? we have a test that seems to contradict that: test_solo @adamgayoso

File: tests/external/test_solo.py (hunk 17 above)
Comment (watiss) [reply tests/external/test_solo.py:10]: there is no functional difference whether or not they do at this time but I can make the change

File: scvi/external/solo/_model.py [hunk 19]
```diff
@@ -182,7 +182,7 @@ def from_scvi_model(
         logger.info("Creating doublets, preparing SOLO model.")
         f = io.StringIO()
         with redirect_stdout(f):
-            setup_anndata(doublet_adata, batch_key=orig_batch_key)
+            _setup_anndata(doublet_adata, batch_key=orig_batch_key)
```
[the remark this answers is not in the document]
Comment (watiss) [reply scvi/external/solo/_model.py:185]: it really depends what the future holds...

File: scvi/external/solo/_model.py [hunk 20]
```diff
@@ -193,7 +193,7 @@ def from_scvi_model(
             doublet_adata.obs[LABELS_KEY] = "doublet"
 
             full_adata = latent_adata.concatenate(doublet_adata)
-            setup_anndata(full_adata, labels_key=LABELS_KEY)
+            _setup_anndata(full_adata, labels_key=LABELS_KEY)
```
[the remark this answers is not in the document]
Comment (watiss) [reply scvi/external/solo/_model.py:196]: @adamgayoso  self or scvi_model? why self?

File: scvi/external/solo/_model.py (hunk 3 above)
Comment (watiss) [reply scvi/external/solo/_model.py:395]: I wasnt asking if we need the copy arg, I was asking if we need the setup_anndata method

File: scvi/data/_anndata.py [hunk 21]
```diff
@@ -70,6 +71,7 @@ def get_from_registry(adata: anndata.AnnData, key: str) -> np.ndarray:
     return data
 
 
+@deprecated(target=None, deprecated_in="0.14.0", remove_in="0.15.0")
```
[the remark this answers is not in the document]
Comment (watiss) [reply scvi/data/_anndata.py]: done

File: scvi/model/_condscvi.py (hunk 6 above)
Comment (adamgayoso) [reply scvi/model/_condscvi.py:210]: in the test it uses `synthetic_iid` which by default runs setup anndata with the labels key, so it's definitely necessary for this model.

Comment (watiss) [reply scvi/model/_condscvi.py:210]: @adamgayoso sure but then it call `setup_anndata` with no labels information, which effectively overwrites whatever ``synthetic_iid`` would have registered. Is the test wrong perhaps?

Comment (jjhong922) [reply scvi/model/_condscvi.py:210]: The test seems wrong if it does that. This method doesn't really do anything if labels don't exist.

File: tests/models/test_models.py (hunk 8 above)
Comment (jjhong922) [reply tests/models/test_models.py:175]: This should still be SCVI.setup_anndata right?

File: tests/models/test_models.py [hunk 22]
```diff
@@ -293,7 +299,7 @@ def test_backed_anndata_scvi(save_path):
     path = os.path.join(save_path, "test_data.h5ad")
     adata.write_h5ad(path)
     adata = anndata.read_h5ad(path, backed="r+")
-    setup_anndata(adata, batch_key="batch")
+    _setup_anndata(adata, batch_key="batch")
```
Comment (jjhong922) [inline tests/models/test_models.py:302]: same here

File: tests/models/test_models.py [hunk 23]
```diff
@@ -696,7 +726,7 @@ def test_multiple_covariates(save_path):
     adata.obs["cont2"] = np.random.normal(size=(adata.shape[0],))
     adata.obs["cat1"] = np.random.randint(0, 5, size=(adata.shape[0],))
     adata.obs["cat2"] = np.random.randint(0, 5, size=(adata.shape[0],))
-    setup_anndata(
+    _setup_anndata(
```
Comment (jjhong922) [inline tests/models/test_models.py:729]: for this one I guess you would need to set it up once with every method, which feels bad. Although I guess we do not expect this scenario to happen often.

**Review by jjhong922 — COMMENTED** [review]
Couple more comments

**Review by jjhong922 — COMMENTED** [review]
Couple more comments

File: scvi/external/solo/_model.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (adamgayoso) [reply scvi/external/solo/_model.py:196]: because `full_adata` corresponds to the input on this model. Practically it shouldn't matter though, but if this model's setup anndata ever adds things, this will be correct

**Comment by adamgayoso** [comment]
Let's make setup anndata an abstract static method as well!

File: scvi/external/gimvi/_model.py (hunk 5 above)
Comment (watiss) [reply scvi/external/gimvi/_model.py]: added in docrep per Adam's suggestion

File: scvi/model/_condscvi.py (hunk 6 above)
Comment (watiss) [reply scvi/model/_condscvi.py]: made those required

File: scvi/data/_anndata.py [hunk 24]
```diff
@@ -70,6 +71,7 @@ def get_from_registry(adata: anndata.AnnData, key: str) -> np.ndarray:
     return data
 
 
+@deprecated(extra="Please use the model-specific setup_anndata methods instead.")
```
Comment (adamgayoso) [inline scvi/data/_anndata.py:74]: should we denote what version we expect to remove it?

Comment (watiss) [reply scvi/data/_anndata.py]: sure, added 0.15.0

**Review by adamgayoso — APPROVED** [review]
